Managed Kyma Migration Update (#461)
kyma-project/istio: hack/managed-kyma-migration/readme.md

If there is no Istio CR, then Lifecycle Manager provisions the default Istio CR defined in the Istio ModuleTemplate. The migration
adds the Istio module to the Kyma CR without overwriting existing module configuration.

## Migration test process

### Test scenarios

Apply the ModuleTemplate for both `fast` and `regular` channels to Dev Control Plane.

#### SAP BTP, Kyma runtime clusters without existing modules

1. Create a Dev SAP BTP, Kyma runtime cluster.
2. Execute the migration.
3. Verify that `istio-controller-manager` is installed and the Istio CR's status is `Ready`.

#### SAP BTP, Kyma runtime cluster with an existing module

1. Create a Dev SAP BTP, Kyma runtime cluster.
2. Add the Keda module to the Kyma CR.
   ```yaml
   spec:
     modules:
       - name: keda
   ```
3. Execute the migration.
4. Verify that `istio-controller-manager` is installed and the Istio CR's status is `Ready`.

## Module's rollout and migration

### Preparations

Executing `kcp taskrun` requires the path to the kubeconfig file of the corresponding Gardener project and permissions to list/get shoots.

### Dev

#### Prerequisites

> - Reconciliation is disabled for the Dev environment. See PR #4485 in the `kyma/management-plane-config` repository.

Reviewer: The PR #4485 belongs to api-gateway. It looks like this is missing for Istio.

adds the Istio module to the Kyma CR without overwriting existing module configuration.

## Migration test process

### Test scenarios

Apply the ModuleTemplate for both `fast` and `regular` channels to Dev Control Plane.

#### SAP BTP, Kyma runtime clusters without existing modules

1. Create a Dev SAP BTP, Kyma runtime cluster.
2. Execute the migration.
3. Verify that `istio-controller-manager` is installed and the Istio CR's status is `Ready`.

#### SAP BTP, Kyma runtime cluster with an existing module

1. Create a Dev SAP BTP, Kyma runtime cluster.
2. Add the Keda module to the Kyma CR.
   ```yaml
   spec:
     modules:
     - name: keda
   ```
3. Execute the migration.
4. Verify that `istio-controller-manager` is installed and the Istio CR's status is `Ready`.

## Module's rollout and migration

### Preparations

Executing `kcp taskrun` requires the path to the kubeconfig file of the corresponding Gardener project and permissions to list/get shoots.

### Dev

#### Migration procedure

1. Push module to `experimental` channel in `kyma/module-manifests `repository (PR #162).

> 2. Push the kustomization change for `experimental` in `kyma/kyma-modules` repository (PR #400)

Reviewer: ```suggestion
2. Push the kustomization change for `experimental` in `kyma/kyma-modules` repository (PR #403)
```

> 3. Verify that the ModuleTemplate is present in the `kyma/kyma-modules` internal repository.

Reviewer: ```suggestion
3. Verify that the ModuleTemplate is present in the experimental channel in the `kyma/kyma-modules` internal repository.
```

> 4. Apply the ModuleTemplate for `fast` and `regular` channels to Dev Control Plane.

Reviewer: ```suggestion
4. Create ModuleTemplate for `fast` and `regular` by adapting the `metadata.name` and `spec.channel` of the `experimental` ModuleTemplate. In this way, we ensure that we also use a ModuleTemplate created by the submission pipeline for the fast and regular channel.
5. Apply the ModuleTemplate for `fast` and `regular` channels to Dev Control Plane.
```

5. Verify that the ModuleTemplate in the `experimental`, `fast` and `regular` channels is available on SAP BTP, Kyma runtime clusters of the Dev environment.

> 6. Merge PR (#4624) in `kyma/management-plane-config` responsible for disabling Istio reconciliation and setting Istio as a default module on dev. Depending on the outcome adjust this those actions to one PR for the stage and prod.

Reviewer: This PR needs to be updated to also include disabling the reconciliation.

> 7. Use `kcp login` to log in to Dev and run the migration script on all SAP BTP, Kyma runtime clusters. To do that, you can use the following command:

Reviewer: ```suggestion
7. Get permissions to execute scripts for `kcp taskrun` and use `kcp login` to log in to Dev and run the migration script on all SAP BTP, Kyma runtime clusters. To do that, you can use the following command:
```

   ```shell
   kcp taskrun --gardener-kubeconfig {PATH TO GARDENER PROJECT KUBECONFIG} -t all -- ./managed-kyma-migration.sh
   ```
8. Verify that the migration worked as expected by checking the status of Istio manifests on Control Plane.
   ```shell
   kubectl get manifests -n kcp-system -o custom-columns=NAME:metadata.name,STATE:status.state | grep istio
   ```

> 9. If there are APIGateway CR is in the warning state enable Istio Module in the Kyma CR (you can check it easily on the new monitoring dashboard).

Reviewer: ```suggestion
9. Verify that no APIGateway module is in Warning state. This can be done by using the [API-Gateway Monitoring Dashboard](https://plutono.cp.dev.kyma.cloud.sap/d/6meO06VSk/modules-api-gateway?orgId=1).
10. If there are APIGateway CR is in the warning state, the cluster might have been created without Istio Module. In this case it needs to be enabled in the Kyma CR manually.
```

   ```shell
   kubectl get manifests -n kcp-system -o custom-columns=NAME:metadata.name,STATE:status.state | grep istio | grep -v Ready
   ```

##### Migration rollout (done by SRE)
1. Apply manually the ModuleTemplate for both `fast` and `regular` channels to Stage Control Plane.
2. Verify that the ModuleTemplate in the `fast` and `regular` channels is available in SAP BTP, Kyma runtime clusters of the Stage environment.
3. Merge PR (#4626) in `kyma/management-plane-config` responsible for disabling Istio reconciliation and setting Istio as a default module
4. Execute `enable-module.sh` script from runtime-ops repository for istio to run default channel migration on all clusters
5. Execute post migration script `post-enable-module.sh` to verify migration.
6. If script failed with following log: `More than one Istio CR present on the cluster. Script rename-to-default.sh might be required`, contact the customer to agree on solution. We propose to execute rename-to-default.sh script.

##### Verify migration

> 1. Wait for SRE to finish the migration for all clusters

Reviewer 2: ```suggestion
1. Wait for SRE to finish the migration for all clusters.
```

> 2. Check the [Istio Module](https://plutono.cp.stage.kyma.cloud.sap/d/hTm72lVIz/modules-istio?orgId=1) and [API Gateway Module](https://plutono.cp.stage.kyma.cloud.sap/d/6meO06VSk/modules-api-gateway?orgId=1) status on the dashboards

Reviewer 2: ```suggestion
2. Check the statuses of the [Istio module](https://plutono.cp.stage.kyma.cloud.sap/d/hTm72lVIz/modules-istio?orgId=1) and the [API Gateway module](https://plutono.cp.stage.kyma.cloud.sap/d/6meO06VSk/modules-api-gateway?orgId=1) on the dashboards.
```

Perform the rollout to Stage together with the SRE team. Since they have already performed the rollout for other modules, they might suggest a different rollout strategy.

#### Prerequisites

- Reconciliation is disabled for the Stage environment. See PR #4486 to the `kyma/management-plane-config` repository.

#### Migration procedure

> 

Reviewer 3: 1. Push the module to experimental `channel` in `kyma/module-manifests` repository.
2. Test that experimental `channel`deploys as expected by manually enabling it on a Stage managed cluster

Perform the rollout to Stage together with the SRE team. Since they have already performed the rollout for other modules, they might suggest a different rollout strategy.

#### Prerequisites

> - Reconciliation is disabled for the Stage environment. See PR #4486 to the `kyma/management-plane-config` repository.

Reviewer: The PR #4486 belongs to api-gateway. It looks like this is missing for Istio.

#### Migration procedure

> 1. Push the module to experimental channel in kyma/module-manifests repository.

Reviewer 3: ```suggestion
1. Push the module to `experimental` channel in `kyma/module-manifests `repository.
```

> 2. Test that experimental channeldeploys as expected by manually enabling it on a Stage managed cluster

Reviewer: ```suggestion
2. Test that experimental channel deploys as expected by manually enabling it on a Stage managed cluster
```

3. Apply the ModuleTemplate for both `fast` and `regular` channels to Stage Control Plane.
4. Verify that the ModuleTemplate in the `fast` and `regular` channels is available in SAP BTP, Kyma runtime clusters of the Stage environment.

> 5. Use `kcp login` to log in to Stage, select a few SAP BTP, Kyma runtime clusters on `Kyma-Test/Kyma-Integration`, and run `managed-kyma-migration.sh` on them using `kcp taskrun`.

Reviewer: What do we do with the cluster from step 2? Shouldn't we change it to the fast channel?

#### Prerequisites

- Reconciliation is disabled for the Stage environment. See PR #4601 to the `kyma/management-plane-config` repository.

#### Migration procedure
1. Push the module to `experimental` channel in `kyma/module-manifests `repository.
2. Test that experimental channel deploys as expected by manually enabling it on a Stage managed cluster
3. Apply the ModuleTemplate for both `fast` and `regular` channels to Stage Control Plane.
4. Verify that the ModuleTemplate in the `fast` and `regular` channels is available in SAP BTP, Kyma runtime clusters of the Stage environment.
5. Use `kcp login` to log in to Stage, select a few SAP BTP, Kyma runtime clusters on `Kyma-Test/Kyma-Integration`, and run `managed-kyma-migration.sh` on them using `kcp taskrun`.
6. Verify if the migration was successful on the SAP BTP, Kyma runtime clusters by checking the status of Istio CR and the reconciler's components.
7. Run `managed-kyma-migration.sh` for all SKRs in `Kyma-Test` and `Kyma-Integration` global accounts.
8. Verify if the migration worked as expected.
9. Run `managed-kyma-migration.sh` for the whole Canary landscape.
10. Verify if the migration worked as expected.

> 

Reviewer: Maybe we should add a step 11. to make sure, executing people from our team remember what to do with failed migrations, e.g. If migration failed with `More than one Istio CR present on the cluster. Script rename-to-default.sh might be required.`, we should contact the customer and propose the execution of the rename-to-default.sh, to fix the Istio CR in error state.
If we want to add this information, we should also add it to the Prod-Steps.
